### PR TITLE
PERF: unstack

### PR DESCRIPTION
**Pull request by jbrockmendel**
```
from asv_bench.benchmarks.reshape import *
self = Unstack()
self.setup("category")

%timeit self.time_without_last_row("category")
77.3 ms ± 1.97 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <- master
21.2 ms ± 788 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <- PR
```

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1692,15 +1695,34 @@ def _unstack(
...
+                type(self)(
+                    self.values.take(
+                        indices, allow_fill=needs_masking[i], fill_value=fill_value
+                    ),
+                    BlockPlacement(place),
+                    ndim=2,
+                )
+                for i, (indices, place) in enumerate(zip(new_values, new_placement))
+            ]
+        else:
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1714]: hmm can you avoid repeating this? (e.g. could create a `needs_masking[i]` with all values == False for the not allow_fill case

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1714]: sure

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -1674,6 +1676,7 @@ def _unstack(
         fill_value,
         new_placement: npt.NDArray[np.intp],
         allow_fill: bool,
+        needs_masking: npt.NDArray[np.bool_],
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1679]: maybe we should now remove allow_fill arg as it doesn't do anything here?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
nice!

**Pull request merged**